Wildcard nodes in parameter file not supported by controller spawner

**Describe the bug**
A clear and concise description of what the bug is.
If the parameter file passed to controller spawner only has wildcard node parameters, spawner fails

**To Reproduce**
Steps to reproduce the behavior:
1. Start controller passing in parameter file to controller spawner with only wildcard node parameter. For example,

```python
load_controller = [
            ExecuteProcess(
                cmd=[
                    "ros2 run controller_manager spawner my_controller "
                    + "--param-file my_controller_params.yaml "
                    + "--controller-manager-timeout 50 "
                    + f"--ros-args --log-level {log_level} --log-file-name my_controller_spawner"
                ],
                name=my_controller_spawner,
                shell=True,
                output=both",
            )
        ]
```
        
        
 my_controller_params.yaml example:
 
```yaml
 /**:
  ros__parameters:
    param1: 1.0
    param2: 2.0
```
    
2. See error like following:

```
[my_controller_spawner-12] [FATAL] [1733946036.807125471] [spawner_my_controller]: Controller : /my_controller parameters not found in parameter file : /home/user/workspace/my_ws/install/app_bringup/share/app_bringup/config/my_controller_params.yaml
[my_controller_spawner-12] Traceback (most recent call last):
[my_controller_spawner-12]   File "/home/user/workspace/my_ws/install/controller_manager/lib/controller_manager/spawner", line 33, in <module>
[my_controller_spawner-12]     sys.exit(load_entry_point('controller-manager==4.20.0', 'console_scripts', 'spawner')())
[my_controller_spawner-12]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[my_controller_spawner-12]   File "/home/user/workspace/my_ws/install/controller_manager/lib/python3.12/site-packages/controller_manager/spawner.py", line 176, in main
[my_controller_spawner-12]     if not set_controller_parameters_from_param_file(
[my_controller_spawner-12]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[my_controller_spawner-12]   File "/home/user/workspace/my_ws/install/controller_manager/lib/python3.12/site-packages/controller_manager/controller_manager_services.py", line 359, in set_controller_parameters_from_param_file
[my_controller_spawner-12]     controller_type = get_parameter_from_param_file(
[my_controller_spawner-12]                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[my_controller_spawner-12]   File "/home/user/workspace/my_ws/install/controller_manager/lib/python3.12/site-packages/controller_manager/controller_manager_services.py", line 303, in get_parameter_from_param_file
[my_controller_spawner-12]     if parameter_name in controller_param_dict[ROS_PARAMS_KEY]:
[my_controller_spawner-12]                          ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
[my_controller_spawner-12] TypeError: 'NoneType' object is not subscriptable
[ERROR] [my_controller_spawner-12]: process has died [pid 62480, exit code 1, cmd 'ros2 run controller_manager spawner my_controller --param-file /home/user/workspace/my_ws/install/app_bringup/share/app_bringup/config/my_controller_params.yaml --controller-manager-timeout 40 --ros-args --log-level ERROR --log-file-name my_controller_spawner'].
```

**Expected behavior**
Controller should start with spawner without any errors

**Screenshots**
N/A

**Environment (please complete the following information):**
 - OS: Ubuntu 24.04
 - Version: rolling sync as of 2024-11-26
 - ros2_control Version: 4.20.0

**Additional context**
Controller starts fine if I have the controller name instead of wildcard. However, I want to use the parameter file with wildcard since I have several instances of this controller and I want to use the same parameter file.

I also tried a test with controller manager parameter lock_memory set to false and it had no change in behavior.

I can't create a minimal reproducible example as my controller is proprietary code and has other dependencies.

--- a/controller_manager/CMakeLists.txt
+++ b/controller_manager/CMakeLists.txt
@@ -222,6 +222,9 @@ if(BUILD_TESTING)
   install(FILES test/test_controller_overriding_parameters.yaml
     DESTINATION test)
 
+  install(FILES test/test_controller_spawner_wildcard_entries.yaml
+    DESTINATION test)
+
   ament_add_gmock(test_hardware_management_srvs
     test/test_hardware_management_srvs.cpp
   )

--- a/controller_manager/controller_manager/controller_manager_services.py
+++ b/controller_manager/controller_manager/controller_manager_services.py
@@ -296,6 +296,7 @@ def get_params_files_with_controller_parameters(
                 f"/{controller_name}" if namespace == "/" else f"{namespace}/{controller_name}"
             )
             WILDCARD_KEY = "/**"
+            ROS_PARAMS_KEY = "ros__parameters"
             parameters = yaml.safe_load(f)
             # check for the parameter in 'controller_name' or 'namespaced_controller' or '/**/namespaced_controller' or '/**/controller_name'
             for key in [
@@ -304,6 +305,8 @@ def get_params_files_with_controller_parameters(
                 f"{WILDCARD_KEY}/{controller_name}",
                 f"{WILDCARD_KEY}{namespaced_controller}",
             ]:
+                if parameter_file in controller_parameter_files:
+                    break
                 if key in parameters:
                     if key == controller_name and namespace != "/":
                         node.get_logger().fatal(
@@ -313,6 +316,8 @@ def get_params_files_with_controller_parameters(
                     controller_parameter_files.append(parameter_file)
 
                 if WILDCARD_KEY in parameters and key in parameters[WILDCARD_KEY]:
+                    controller_parameter_files.append(parameter_file)
+                if WILDCARD_KEY in parameters and ROS_PARAMS_KEY in parameters[WILDCARD_KEY]:
                     controller_parameter_files.append(parameter_file)
     return controller_parameter_files
 
@@ -346,6 +351,8 @@ def get_parameter_from_param_files(
 
                 if WILDCARD_KEY in parameters and key in parameters[WILDCARD_KEY]:
                     controller_param_dict = parameters[WILDCARD_KEY][key]
+                if WILDCARD_KEY in parameters and ROS_PARAMS_KEY in parameters[WILDCARD_KEY]:
+                    controller_param_dict = parameters[WILDCARD_KEY]
 
                 if controller_param_dict and (
                     not isinstance(controller_param_dict, dict)

--- a/controller_manager/doc/userdoc.rst
+++ b/controller_manager/doc/userdoc.rst
@@ -185,7 +185,7 @@ The parsed controller config file can follow the same conventions as the typical
       ros__parameters:
         joints:
           - joint1
-    
+
     position_trajectory_controller_joint2:
       ros__parameters:
         joints:

--- a/controller_manager/doc/userdoc.rst
+++ b/controller_manager/doc/userdoc.rst
@@ -173,55 +173,75 @@ The parsed controller config file can follow the same conventions as the typical
 
  .. code-block:: yaml
 
-    /**/position_trajectory_controller:
-    ros__parameters:
-      type: joint_trajectory_controller/JointTrajectoryController
-      joints:
-        - joint1
-        - joint2
+    /**:
+      ros__parameters:
+        type: joint_trajectory_controller/JointTrajectoryController
 
-      command_interfaces:
-        - position
-        .....
+        command_interfaces:
+          - position
+          .....
+
+    position_trajectory_controller_joint1:
+      ros__parameters:
+        joints:
+          - joint1
+    
+    position_trajectory_controller_joint2:
+      ros__parameters:
+        joints:
+          - joint2
+
+ .. code-block:: yaml
+
+    /**/position_trajectory_controller:
+      ros__parameters:
+        type: joint_trajectory_controller/JointTrajectoryController
+        joints:
+          - joint1
+          - joint2
+
+        command_interfaces:
+          - position
+          .....
 
  .. code-block:: yaml
 
     /position_trajectory_controller:
-    ros__parameters:
-      type: joint_trajectory_controller/JointTrajectoryController
-      joints:
-        - joint1
-        - joint2
+      ros__parameters:
+        type: joint_trajectory_controller/JointTrajectoryController
+        joints:
+          - joint1
+          - joint2
 
-      command_interfaces:
-        - position
-        .....
+        command_interfaces:
+          - position
+          .....
 
  .. code-block:: yaml
 
     position_trajectory_controller:
-    ros__parameters:
-      type: joint_trajectory_controller/JointTrajectoryController
-      joints:
-        - joint1
-        - joint2
+      ros__parameters:
+        type: joint_trajectory_controller/JointTrajectoryController
+        joints:
+          - joint1
+          - joint2
 
-      command_interfaces:
-        - position
-        .....
+        command_interfaces:
+          - position
+          .....
 
  .. code-block:: yaml
 
     /rrbot_1/position_trajectory_controller:
-    ros__parameters:
-      type: joint_trajectory_controller/JointTrajectoryController
-      joints:
-        - joint1
-        - joint2
+      ros__parameters:
+        type: joint_trajectory_controller/JointTrajectoryController
+        joints:
+          - joint1
+          - joint2
 
-      command_interfaces:
-        - position
-        .....
+        command_interfaces:
+          - position
+          .....
 
 ``unspawner``
 ^^^^^^^^^^^^^^^^

--- a/controller_manager/test/test_controller_spawner_wildcard_entries.yaml
+++ b/controller_manager/test/test_controller_spawner_wildcard_entries.yaml
@@ -1,0 +1,6 @@
+/**:
+  ros__parameters:
+    type: "controller_manager/test_controller"
+    joint_names: ["joint1"]
+    param1: 1.0
+    param2: 2.0

--- a/controller_manager/test/test_controller_spawner_wildcard_entries.yaml
+++ b/controller_manager/test/test_controller_spawner_wildcard_entries.yaml
@@ -4,3 +4,7 @@
     joint_names: ["joint1"]
     param1: 1.0
     param2: 2.0
+
+wildcard_ctrl_3:
+  ros__parameters:
+    param3: 3.0

--- a/controller_manager/test/test_spawner_unspawner.cpp
+++ b/controller_manager/test/test_spawner_unspawner.cpp
@@ -366,6 +366,62 @@ TEST_F(TestLoadController, spawner_test_type_in_params_file)
     test_file_path);
 }
 
+TEST_F(TestLoadController, spawner_test_with_wildcard_entries_with_no_ctrl_name)
+{
+  const std::string test_file_path = ament_index_cpp::get_package_prefix("controller_manager") +
+                                     "/test/test_controller_spawner_wildcard_entries.yaml";
+
+  ControllerManagerRunner cm_runner(this);
+  // Provide controller type via the parsed file
+  EXPECT_EQ(
+    call_spawner(
+      "wildcard_ctrl_1 wildcard_ctrl_2 -c test_controller_manager --controller-manager-timeout 1.0 "
+      "-p " +
+      test_file_path),
+    0);
+
+  auto verify_ctrl_parameter = [](const auto & ctrl_node)
+  {
+    if (!ctrl_node->has_parameter("joint_names"))
+    {
+      ctrl_node->declare_parameter("joint_names", std::vector<std::string>({"random_joint"}));
+    }
+    ASSERT_THAT(
+      ctrl_node->get_parameter("joint_names").as_string_array(),
+      std::vector<std::string>({"joint1"}));
+
+    if (!ctrl_node->has_parameter("param1"))
+    {
+      ctrl_node->declare_parameter("param1", -10.0);
+    }
+    ASSERT_THAT(ctrl_node->get_parameter("param1").as_double(), 1.0);
+
+    if (!ctrl_node->has_parameter("param2"))
+    {
+      ctrl_node->declare_parameter("param2", -10.0);
+    }
+    ASSERT_THAT(ctrl_node->get_parameter("param2").as_double(), 2.0);
+  };
+
+  ASSERT_EQ(cm_->get_loaded_controllers().size(), 2ul);
+
+  auto wildcard_ctrl_1 = cm_->get_loaded_controllers()[0];
+  ASSERT_EQ(wildcard_ctrl_1.info.name, "wildcard_ctrl_2");
+  ASSERT_EQ(wildcard_ctrl_1.info.type, test_controller::TEST_CONTROLLER_CLASS_NAME);
+  ASSERT_EQ(
+    wildcard_ctrl_1.c->get_lifecycle_state().id(),
+    lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE);
+  verify_ctrl_parameter(wildcard_ctrl_1.c->get_node());
+
+  auto wildcard_ctrl_2 = cm_->get_loaded_controllers()[1];
+  ASSERT_EQ(wildcard_ctrl_2.info.name, "wildcard_ctrl_1");
+  ASSERT_EQ(wildcard_ctrl_2.info.type, test_controller::TEST_CONTROLLER_CLASS_NAME);
+  ASSERT_EQ(
+    wildcard_ctrl_2.c->get_lifecycle_state().id(),
+    lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE);
+  verify_ctrl_parameter(wildcard_ctrl_2.c->get_node());
+}
+
 TEST_F(TestLoadController, unload_on_kill)
 {
   // Launch spawner with unload on kill

--- a/controller_manager/test/test_spawner_unspawner.cpp
+++ b/controller_manager/test/test_spawner_unspawner.cpp
@@ -375,12 +375,13 @@ TEST_F(TestLoadController, spawner_test_with_wildcard_entries_with_no_ctrl_name)
   // Provide controller type via the parsed file
   EXPECT_EQ(
     call_spawner(
-      "wildcard_ctrl_1 wildcard_ctrl_2 -c test_controller_manager --controller-manager-timeout 1.0 "
+      "wildcard_ctrl_1 wildcard_ctrl_2 wildcard_ctrl_3 -c test_controller_manager "
+      "--controller-manager-timeout 1.0 "
       "-p " +
       test_file_path),
     0);
 
-  auto verify_ctrl_parameter = [](const auto & ctrl_node)
+  auto verify_ctrl_parameter = [](const auto & ctrl_node, bool has_param_3)
   {
     if (!ctrl_node->has_parameter("joint_names"))
     {
@@ -401,25 +402,39 @@ TEST_F(TestLoadController, spawner_test_with_wildcard_entries_with_no_ctrl_name)
       ctrl_node->declare_parameter("param2", -10.0);
     }
     ASSERT_THAT(ctrl_node->get_parameter("param2").as_double(), 2.0);
+
+    if (!ctrl_node->has_parameter("param3"))
+    {
+      ctrl_node->declare_parameter("param3", -10.0);
+    }
+    ASSERT_THAT(ctrl_node->get_parameter("param3").as_double(), has_param_3 ? 3.0 : -10.0);
   };
 
-  ASSERT_EQ(cm_->get_loaded_controllers().size(), 2ul);
+  ASSERT_EQ(cm_->get_loaded_controllers().size(), 3ul);
 
-  auto wildcard_ctrl_1 = cm_->get_loaded_controllers()[0];
-  ASSERT_EQ(wildcard_ctrl_1.info.name, "wildcard_ctrl_2");
-  ASSERT_EQ(wildcard_ctrl_1.info.type, test_controller::TEST_CONTROLLER_CLASS_NAME);
+  auto wildcard_ctrl_3 = cm_->get_loaded_controllers()[0];
+  ASSERT_EQ(wildcard_ctrl_3.info.name, "wildcard_ctrl_3");
+  ASSERT_EQ(wildcard_ctrl_3.info.type, test_controller::TEST_CONTROLLER_CLASS_NAME);
   ASSERT_EQ(
-    wildcard_ctrl_1.c->get_lifecycle_state().id(),
+    wildcard_ctrl_3.c->get_lifecycle_state().id(),
     lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE);
-  verify_ctrl_parameter(wildcard_ctrl_1.c->get_node());
+  verify_ctrl_parameter(wildcard_ctrl_3.c->get_node(), true);
 
   auto wildcard_ctrl_2 = cm_->get_loaded_controllers()[1];
-  ASSERT_EQ(wildcard_ctrl_2.info.name, "wildcard_ctrl_1");
+  ASSERT_EQ(wildcard_ctrl_2.info.name, "wildcard_ctrl_2");
   ASSERT_EQ(wildcard_ctrl_2.info.type, test_controller::TEST_CONTROLLER_CLASS_NAME);
   ASSERT_EQ(
     wildcard_ctrl_2.c->get_lifecycle_state().id(),
     lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE);
-  verify_ctrl_parameter(wildcard_ctrl_2.c->get_node());
+  verify_ctrl_parameter(wildcard_ctrl_2.c->get_node(), false);
+
+  auto wildcard_ctrl_1 = cm_->get_loaded_controllers()[2];
+  ASSERT_EQ(wildcard_ctrl_1.info.name, "wildcard_ctrl_1");
+  ASSERT_EQ(wildcard_ctrl_1.info.type, test_controller::TEST_CONTROLLER_CLASS_NAME);
+  ASSERT_EQ(
+    wildcard_ctrl_1.c->get_lifecycle_state().id(),
+    lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE);
+  verify_ctrl_parameter(wildcard_ctrl_1.c->get_node(), false);
 }
 
 TEST_F(TestLoadController, unload_on_kill)

